TestBlockTracker: panic: test timed out after 10m0s

https://github.com/erigontech/erigon/actions/runs/15163194948/job/42634093045?pr=14994

```
=== RUN   TestBlockTracker
=== PAUSE TestBlockTracker
...
=== CONT  TestBlockTracker
...
=== NAME  TestBlockTracker
    block_listener.go:55: INFO[05-21|13:33:13.759] running block listener 
    block_tracker.go:50: INFO[05-21|13:33:13.809] running block tracker 
    block_tracker.go:65: DBUG[05-21|13:33:13.809] block tracker setting initial block num  blockNum=10
...
=== NAME  TestBlockTracker
    block_tracker_test.go:60: 
        	Error Trace:	github.com/erigontech/erigon/txnprovider/shutter/block_tracker_test.go:60
        	Error:      	Received unexpected error:
        	            	context deadline exceeded
        	Test:       	TestBlockTracker
panic: test timed out after 10m0s
	running tests:
		TestBlockTracker (10m0s)

goroutine 104 [running]:
testing.(*M).startAlarm.func1()
	testing/testing.go:2373 +0x385
created by time.goFunc
	time/sleep.go:215 +0x2d

goroutine 1 [chan receive, 10 minutes]:
testing.tRunner.func1()
	testing/testing.go:1651 +0x489
testing.tRunner(0xc00092f040, 0xc001037c70)
	testing/testing.go:1696 +0x104
testing.runTests(0xc00116a7c8, {0x7ff74c17f2a0, 0x9, 0x9}, {0x7ff748eb5bd0?, 0x7ff748eb583a?, 0x7ff74eaa5ba0?})
	testing/testing.go:2166 +0x43d
testing.(*M).Run(0xc000953ae0)
	testing/testing.go:2034 +0x64a
main.main()
	_testmain.go:63 +0x9b

goroutine 13 [semacquire, 10 minutes]:
sync.runtime_Semacquire(0x0?)
	runtime/sema.go:71 +0x25
sync.(*WaitGroup).Wait(0xc000521c01?)
	sync/waitgroup.go:118 +0x48
golang.org/x/sync/errgroup.(*Group).Wait(0xc00115f740)
	golang.org/x/sync@v0.14.0/errgroup/errgroup.go:65 +0x25
runtime.Goexit()
	runtime/panic.go:635 +0x5e
testing.(*common).FailNow(0xc00092f6c0)
	testing/testing.go:1006 +0x4a
github.com/stretchr/testify/require.NoError({0x7ff74d18d078, 0xc00092f6c0}, {0x7ff74d17ee40, 0x7ff74eaf9e60}, {0x0, 0x0, 0x0})
	github.com/stretchr/testify@v1.10.0/require/require.go:1357 +0xca
github.com/erigontech/erigon/txnprovider/shutter_test.TestBlockTracker(0xc00092f6c0)
	github.com/erigontech/erigon/txnprovider/shutter/block_tracker_test.go:60 +0x5e7
testing.tRunner(0xc00092f6c0, 0x7ff74ce36b28)
	testing/testing.go:1690 +0xcb
created by testing.(*T).Run in goroutine 1
	testing/testing.go:1743 +0x377

goroutine 68 [select, 10 minutes]:
github.com/erigontech/erigon/txnprovider/shutter/internal/testhelpers.kvStateChangesStreamMock.Recv({{0x7ff74d19ef00?, 0xc000a15540?}, 0xc0000570a0?})
	github.com/erigontech/erigon/txnprovider/shutter/internal/testhelpers/state_changes_client_mock.go:50 +0x8e
github.com/erigontech/erigon/txnprovider/shutter.(*BlockListener).Run(0xc001414a20, {0x7ff74d19ef00, 0xc000a15590})
	github.com/erigontech/erigon/txnprovider/shutter/block_listener.go:64 +0x13c
github.com/erigontech/erigon/txnprovider/shutter_test.TestBlockTracker.func2()
	github.com/erigontech/erigon/txnprovider/shutter/block_tracker_test.go:48 +0x1f
golang.org/x/sync/errgroup.(*Group).add.func1()
	golang.org/x/sync@v0.14.0/errgroup/errgroup.go:130 +0x7e
created by golang.org/x/sync/errgroup.(*Group).add in goroutine 13
	golang.org/x/sync@v0.14.0/errgroup/errgroup.go:98 +0x79

goroutine 69 [select, 10 minutes]:
github.com/erigontech/erigon/txnprovider/shutter.(*BlockTracker).Run(0xc001432380, {0x7ff74d19ef00, 0xc000a15590})
	github.com/erigontech/erigon/txnprovider/shutter/block_tracker.go:82 +0x3e5
github.com/erigontech/erigon/txnprovider/shutter_test.TestBlockTracker.func3()
	github.com/erigontech/erigon/txnprovider/shutter/block_tracker_test.go:49 +0x1f
golang.org/x/sync/errgroup.(*Group).add.func1()
	golang.org/x/sync@v0.14.0/errgroup/errgroup.go:130 +0x7e
created by golang.org/x/sync/errgroup.(*Group).add in goroutine 13
	golang.org/x/sync@v0.14.0/errgroup/errgroup.go:98 +0x79
FAIL	github.com/erigontech/erigon/txnprovider/shutter	600.619s
```

--- a/txnprovider/shutter/block_tracker.go
+++ b/txnprovider/shutter/block_tracker.go
@@ -57,6 +57,15 @@ func (bt *BlockTracker) Run(ctx context.Context) error {
 		bt.blockChangeCond.L.Unlock()
 	}()
 
+	blockEventC := make(chan BlockEvent)
+	unregisterBlockEventObserver := bt.blockListener.RegisterObserver(func(blockEvent BlockEvent) {
+		select {
+		case <-ctx.Done(): // no-op
+		case blockEventC <- blockEvent:
+		}
+	})
+	defer unregisterBlockEventObserver()
+
 	bn, err := bt.currentBlockNumReader(ctx)
 	if err != nil {
 		return err
@@ -69,15 +78,6 @@ func (bt *BlockTracker) Run(ctx context.Context) error {
 		bt.blockChangeCond.L.Unlock()
 	}
 
-	blockEventC := make(chan BlockEvent)
-	unregisterBlockEventObserver := bt.blockListener.RegisterObserver(func(blockEvent BlockEvent) {
-		select {
-		case <-ctx.Done(): // no-op
-		case blockEventC <- blockEvent:
-		}
-	})
-
-	defer unregisterBlockEventObserver()
 	for {
 		select {
 		case <-ctx.Done():

--- a/txnprovider/shutter/block_tracker_test.go
+++ b/txnprovider/shutter/block_tracker_test.go
@@ -18,6 +18,7 @@ package shutter_test
 
 import (
 	"context"
+	"sync"
 	"testing"
 	"time"
 
@@ -38,8 +39,11 @@ func TestBlockTracker(t *testing.T) {
 	logger := testlog.Logger(t, log.LvlTrace)
 	recvC := make(chan *remoteproto.StateChangeBatch)
 	bl := shutter.NewBlockListener(logger, testhelpers.NewStateChangesClientMock(ctx, recvC))
+	btInitialisationWg := &sync.WaitGroup{}
+	btInitialisationWg.Add(1)
 	bnReader := func(ctx context.Context) (*uint64, error) {
 		start := uint64(10)
+		btInitialisationWg.Done()
 		return &start, nil
 	}
 	bt := shutter.NewBlockTracker(logger, bl, bnReader)
@@ -51,6 +55,7 @@ func TestBlockTracker(t *testing.T) {
 	}(eg)
 	eg.Go(func() error { return bl.Run(egCtx) })
 	eg.Go(func() error { return bt.Run(egCtx) })
+	btInitialisationWg.Wait() // This ensures the block tracker has been initialised i.e. it observes the block listener
 
 	waitCtx1, waitCtxCancel1 := context.WithTimeout(ctx, 50*time.Millisecond)
 	defer waitCtxCancel1()

--- a/txnprovider/shutter/block_tracker_test.go
+++ b/txnprovider/shutter/block_tracker_test.go
@@ -44,7 +44,11 @@ func TestBlockTracker(t *testing.T) {
 	}
 	bt := shutter.NewBlockTracker(logger, bl, bnReader)
 	eg, egCtx := errgroup.WithContext(ctx)
-	defer eg.Wait()
+	defer func(eg *errgroup.Group) {
+		cancel()
+		err := eg.Wait()
+		require.ErrorIs(t, err, context.Canceled)
+	}(eg)
 	eg.Go(func() error { return bl.Run(egCtx) })
 	eg.Go(func() error { return bt.Run(egCtx) })
 
@@ -58,8 +62,4 @@ func TestBlockTracker(t *testing.T) {
 	defer waitCtxCancel2()
 	err = bt.Wait(waitCtx2, 15)
 	require.NoError(t, err)
-
-	cancel()
-	err = eg.Wait()
-	require.ErrorIs(t, err, context.Canceled)
 }

